src/smartcontracts/codec/managedDecimal.ts

```diff
@@ -15,35 +16,32 @@ export class ManagedDecimalCodec {
         const payload = buffer.slice(0, length);
 
         const result = this.decodeTopLevel(payload, type);
-        const decodedLength = length;
-
-        return [result, decodedLength];
+        return [result, length];
     }
 
     decodeTopLevel(buffer: Buffer, type: ManagedDecimalType): ManagedDecimalValue {
         if (buffer.length === 0) {
-            return new ManagedDecimalValue(new BigNumber(0), 2);
+            return new ManagedDecimalValue(new BigNumber(0), 0);
         }
 
-        const isUsize = type.getMetadata() === "usize";
-
-        if (isUsize) {
-            const u32Size = 4;
-            const bigUintSize = buffer.length - u32Size;
+        if (type.isVariable()) {
+            const bigUintSize = buffer.length - SizeOfU32;
 
-            const bigUint = new BigNumber(buffer.slice(0, bigUintSize).toString("hex"), 16);
-            const u32 = buffer.readUInt32BE(bigUintSize);
+            const value = new BigNumber(buffer.slice(0, bigUintSize).toString("hex"), 16);
+            const scale = buffer.readUInt32BE(bigUintSize);
 
-            return new ManagedDecimalValue(bigUint, u32);
+            return new ManagedDecimalValue(value, scale);
         }
 
         const value = bufferToBigInt(buffer);
-        return new ManagedDecimalValue(value, parseInt(type.getMetadata()));
+        const scale = parseInt(type.getMetadata());
+        return new ManagedDecimalValue(value, scale);
     }
 
     encodeNested(value: ManagedDecimalValue): Buffer {
         let buffers: Buffer[] = [];
-        if (value.getType().getMetadata() == "usize") {
+        const managedDecimalType = value.getType() as ManagedDecimalType;
```

Comment:
If we add `isVariable()` not only to the type, but also to the value, we work around without type assertions.